src/main/java/Tree/WeightBalancedTree.java

```diff
@@ -0,0 +1,170 @@
+package Tree;
+
+import Node.WeightBalancedNode;
+
+
+public class WeightBalancedTree<T> extends BinarySearchTree<T> {
+
+    public WeightBalancedTree() { current = null; }
+
+    public WeightBalancedTree(int key, T value, int weight) {
+        current = new WeightBalancedNode<T>(key, value, weight);
+        root = current;
+    }
+
+    public void insert(int insert_key, T value, int weight) {
+        current = insertKey(insert_key, value, weight, (WeightBalancedNode<T>) root);
+        if (current.getParentNode() == null) {
+            root = current;
+        }
+    }
+
+    private WeightBalancedNode<T> insertKey(int insert_key, T value, int weight, WeightBalancedNode<T> node) {
+        if (node == null) { return null; }
+        WeightBalancedNode<T> right = (WeightBalancedNode) node.getRightNode();
+        WeightBalancedNode<T> left = (WeightBalancedNode) node.getLeftNode();
+
+        if (node.getKey() < insert_key) {
+            if (right == null) {
+                WeightBalancedNode<T> newNode = new WeightBalancedNode(insert_key, value, weight);
+                linkNodes(node, newNode, 'R');
+            } else {
+                insertKey(insert_key, value, weight, right);
+            }
+            if (((WeightBalancedNode<T>) node.getRightNode()).getWeight() > node.getWeight()) {
```

Comment:
Can't you move this line to `if (right ==null)` block and use newNode?

Reply:
then the rotation will only occur once at the leaf insertion